### DISCUSSION TITLE
Support for Multiple Source and Target Positions?

### DISCUSSION THREAD

**Discussion by pjatx** [bcakmakoglu/vue-flow#270] [Q&A] [answered]
Hey There, 

Thanks for the amazing library! Wondering if it's possible to allow edges to connect to more than one position / side of a node. I have a complex workflow where I want nodes coming in and exiting from both the tops / bottoms and sides and am unsure how to accomplish that currently. 

Any help would be greatly appreciated.

**Comment by bcakmakoglu** [comment, marked as answer]
You can implement custom nodes that have as many handles as you want on whatever side you want. 
Check the [Custom Node Example](https://vueflow.dev/examples/nodes/). In that example there is a node with 2 handles on the right side.
Also read [this part](https://vueflow.dev/guide/node.html#dynamic-handle-positions-adding-handles-dynamically) of the docs if you plan on adding handles dynamically.

If you want edges to be able to fork, you'll have to implement your own custom edges as the built-in ones only handle connection from one handle to another.

**Reply by pjatx** [reply]
Got it, makes sense. Thank you for the quick reply : )

**Reply by pjatx** [reply]
@bcakmakoglu I am actually wanting to add handles dynamically or rather based on the data object of a custom node. IE 

```
// Node Definition
  {
    id: "4",
    label: "My Label",
    type: "NodeSegment",
    data: {
      handles: ["Left", "Top"],
    },
    position: { x: 0, y: 250 },
    sourcePosition: Position.Right,
    targetPosition: Position.Left,
  },
  ```
  
```
// Custom Node 
<script setup>
import { Handle, Position, useVueFlow } from '@braks/vue-flow'

const props = defineProps({
  data: {
    type: Object,
    required: true
  },
  label: {
    type: String,
    required: true
  }
})

const onConnect = (params) => console.log('handle onConnect', params)
</script>

<template>
  <div class="vue-flow__node-default">
    <div>{{ label }}</div>
  </div>

  <Handle
    v-if="data"
    v-for="handle in data.handles"
    type="source"
    :position="'Position.' + handle"
    :id="handle.toLowerCase()"
  />

</template>
  ```  
  
  Is something like that possible? Have also tried using the `updateNodeInternals` method and can't seem to get it to work.

**Reply by pjatx** [reply]
To clarify, no problem adding the handles like this, but I get errors when trying to add the edges (even after mount).

**Reply by bcakmakoglu** [reply]
` :position="'Position.' + handle"` what's this supposed to be?
Position has to be of type `Position` not some string of `'Position.Left'` but actually `Position.Left`
In your case you could just do `Position[direction]` instead.

Also something about Vue in general: Do not use a `v-if` on a `v-for` (this should cause a warning in your IDE anyway) - use a computed property or something to filter the condition or assign a default value to `data.handles` so you don't need to check the `v-if`.

Other than that just make sure to call `updateNodeInternals` or emit that event when adding new handles ([see this part of the docs](https://vueflow.dev/guide/node.html#dynamic-handle-positions-adding-handles-dynamically)).

**Reply by pjatx** [reply]
Ok thanks - yeah, the handles were just rendering in a weird way when enclosing in a div - even if I make the div fill it's parent container, but will keep playing around with this.

**Reply by bcakmakoglu** [reply]
Can you give me a screenshot of what you mean?

**Reply by pjatx** [reply]
Yeah, sure thing. I'll just setup a sandbox and see if I can reproduce. 

Other weird thing is that I have tried default values and they still return undefined. Still getting up to speed with the composition API after being mostly in Vue 2.

**Reply by bcakmakoglu** [reply]
```vue
const props = withDefaults(
  defineProps({
    data: {
      type: Object,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
  }),
  {
    data: () => ({}),
  },
)
```

This would be how you define defaults for `defineProps` in composition API.
If you want to destructure props (similar to React) you'll have to enable [`reactivityTransform`](https://vuejs.org/guide/extras/reactivity-transform.html).

Then you could just do this
```vue
<script setup>
const { data = {} } = defineProps({
  data: {
    type: Object,
    required: true,
  },
  label: {
    type: String,
    required: true,
  },
})
</script>
```

But without `reactivityTransform` destructuring props makes them lose their reactivity.

**Reply by pjatx** [reply]
Ah that makes sense. I actually realized I was accessing the props incorrectly - thought they would be available under the `data` key but they are in `props.data` instead. Unstuck now and no more `v-if`. Thank you : )
